File: app/src/main/java/com/attendance/tracker/data/model/Subject.kt
package com.attendance.tracker.data.model

import androidx.room.Entity
import androidx.room.PrimaryKey

/**
 * Represents a subject/course for attendance tracking
 */
@Entity(tableName = "subjects")
data class Subject(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val name: String,
    val requiredAttendance: Int = 75, // Default 75% required attendance
    val totalLectures: Int = 0,
    val presentLectures: Int = 0,
    val absentLectures: Int = 0,
    val parentSubjectId: Long? = null, // Null for top-level subjects, or ID of parent subject for sub-subjects
    val isFolder: Boolean = false // True if this is a folder/group (like "Pathology"), false for actual subjects (like "Pathology - Lecture")
) {
    val currentAttendancePercentage: Float
        get() = if (totalLectures > 0) (presentLectures.toFloat() / totalLectures) * 100 else 0f
    
    val isAboveRequired: Boolean
        get() = currentAttendancePercentage >= requiredAttendance
    
    /**
     * Calculate how many classes need to be attended to meet the required attendance
     * Returns 0 if already above required attendance
     */
    val classesToAttend: Int
        get() {
            if (isAboveRequired || totalLectures == 0) return 0
            
            // Formula: (P + x) / (T + x) = R/100
            // Where P = present, T = total, R = required%, x = classes to attend
            // Solving: x = (R*T - 100*P) / (100 - R)
            val numerator = (requiredAttendance * totalLectures) - (100 * presentLectures)
            val denominator = 100 - requiredAttendance
            
            return if (denominator > 0) {
                kotlin.math.ceil(numerator.toDouble() / denominator).toInt().coerceAtLeast(0)
            } else {
                0
            }
        }
    
    /**
     * Calculate how many classes can be bunked while maintaining required attendance
     * Returns 0 if below required attendance or at the threshold
     */
    val classesCanBunk: Int
        get() {
            if (!isAboveRequired || totalLectures == 0) return 0
            
            // Formula: (P) / (T + x) = R/100
            // Where P = present, T = total, R = required%, x = classes can bunk
            // Solving: x = (100*P/R) - T
            val maxTotalWithBunks = (100.0 * presentLectures) / requiredAttendance
            val canBunk = kotlin.math.floor(maxTotalWithBunks - totalLectures).toInt()
            
            return canBunk.coerceAtLeast(0)
        }
}

/**
 * Get the display name for a subject, including folder name if applicable
 */
fun Subject.getDisplayName(allSubjects: List<Subject>): String {
    return if (parentSubjectId != null) {
        val folder = allSubjects.find { it.id == parentSubjectId }
        if (folder != null) {
            "${folder.name} / $name"
        } else {
            name
        }
    } else {
        name
    }
}

File: iOS/AttendanceTracker/AttendanceTracker/Models/Subject.swift
import Foundation
import SwiftData

/// Represents a subject/course for attendance tracking
@Model
final class Subject {
    @Attribute(.unique) var id: UUID
    var name: String
    var requiredAttendance: Int
    var totalLectures: Int
    var presentLectures: Int
    var absentLectures: Int
    var parentSubjectId: UUID?
    var isFolder: Bool
    
    init(
        id: UUID = UUID(),
        name: String,
        requiredAttendance: Int = 75,
        totalLectures: Int = 0,
        presentLectures: Int = 0,
        absentLectures: Int = 0,
        parentSubjectId: UUID? = nil,
        isFolder: Bool = false
    ) {
        self.id = id
        self.name = name
        self.requiredAttendance = requiredAttendance
        self.totalLectures = totalLectures
        self.presentLectures = presentLectures
        self.absentLectures = absentLectures
        self.parentSubjectId = parentSubjectId
        self.isFolder = isFolder
    }
    
    var currentAttendancePercentage: Float {
        guard totalLectures > 0 else { return 0 }
        return (Float(presentLectures) / Float(totalLectures)) * 100
    }
    
    var isAboveRequired: Bool {
        currentAttendancePercentage >= Float(requiredAttendance)
    }
    
    /// Get the display name for a subject, including folder name if applicable
    func getDisplayName(allSubjects: [Subject]) -> String {
        if let parentId = parentSubjectId,
           let folder = allSubjects.first(where: { $0.id == parentId }) {
            return "\(folder.name) / \(name)"
        }
        return name
    }
}
